pkg/internal/storage/repo/sync_state.go

```diff
@@ -0,0 +1,123 @@
+package repo
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/satoshi"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/storage/database/models"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/storage/database/scopes"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/storage/entity"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/wdk"
+	"github.com/go-softwarelab/common/pkg/to"
+	"gorm.io/gorm"
+)
+
+type SyncState struct {
+	db *gorm.DB
+}
+
+func NewSyncState(db *gorm.DB) *SyncState {
+	return &SyncState{
+		db: db,
+	}
+}
+
+func (s *SyncState) FindSyncState(ctx context.Context, userID int, storageIdentityKey string) (*entity.SyncState, error) {
+	var model models.SyncState
+	err := s.db.WithContext(ctx).
+		Scopes(scopes.UserID(userID)).
+		Where("storage_identity_key = ?", storageIdentityKey).
+		First(&model).Error
+
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed to find sync state: %w", err)
+	}
+
+	return mapModelToSyncStateEntity(model)
+}
+
+func (s *SyncState) CreateSyncState(ctx context.Context, syncState *entity.SyncState) (*entity.SyncState, error) {
+	syncMapJSON, err := syncState.SyncMap.JSON()
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal sync map: %w", err)
+	}
+
+	model := models.SyncState{
+		UserID:             syncState.UserID,
+		StorageIdentityKey: syncState.StorageIdentityKey,
+		StorageName:        syncState.StorageName,
+		Status:             syncState.Status,
+		RefNum:             syncState.Reference,
+		SyncMap:            syncMapJSON,
+		When:               syncState.When,
+	}
+
+	if syncState.Satoshis != nil {
+		model.Satoshis = to.Ptr(syncState.Satoshis.Int64())
+	}
+
+	err = s.db.WithContext(ctx).Create(&model).Error
```

Comment:
```suggestion
	err = s.db.WithContext(ctx).Clauses(clause.OnConflict{
        Columns:   []clause.Column{{Name: "user_id"}, {Name: "storage_identity_key"}},
        DoUpdates: clause.AssignmentColumns([]string{"status", "sync_map", "when", "ref_num", "storage_name", "satoshis"}),
    }).Create(&model).Error
```
would not be better ? this way u would  not need `FindSyncState` ?

Reply:
The `Find` is needed anyway because this functions should:
- return existing row if exists
- create if doesn't exist

"Upsert" (the `INSERT` with `ON CONFLICT`) doesn't do the same job.